shared-bindings/gamepad/GamePad.c

```diff
@@ -96,23 +105,86 @@
 //|
 STATIC mp_obj_t gamepad_make_new(const mp_obj_type_t *type, size_t n_args,
         const mp_obj_t *args, mp_map_t *kw_args) {
-    if (n_args > 8) {
-        mp_raise_TypeError(translate("too many arguments"));
+    if (n_args > 8 || n_args == 0) {
+        mp_raise_TypeError(translate("argument num/types mismatch"));
     }
     for (size_t i = 0; i < n_args; ++i) {
-        if (!MP_OBJ_IS_TYPE(args[i], &digitalio_digitalinout_type)) {
-            mp_raise_TypeError(translate("expected a DigitalInOut"));
-        }
-        digitalio_digitalinout_obj_t *pin = MP_OBJ_TO_PTR(args[i]);
-        raise_error_if_deinited(
-            common_hal_digitalio_digitalinout_deinited(pin));
+        validate_pin(args[i]);
     }
-    if (!MP_STATE_VM(gamepad_singleton)) {
-        gamepad_obj_t* gamepad_singleton = m_new_obj(gamepad_obj_t);
+    gamepad_obj_t* gamepad_singleton = MP_STATE_VM(gamepad_singleton);
+    if (!gamepad_singleton) {
+        gamepad_singleton = m_new_obj(gamepad_obj_t);
         gamepad_singleton->base.type = &gamepad_type;
-        MP_STATE_VM(gamepad_singleton) = gc_make_long_lived(gamepad_singleton);
+        gamepad_singleton = gc_make_long_lived(gamepad_singleton);
+        MP_STATE_VM(gamepad_singleton) = gamepad_singleton;
+    }
+    for (size_t i = 0; i < 8; ++i) {
+        gamepad_singleton->pins[i] = NULL;
+    }
+    gamepad_singleton->pulls = 0;
+    for (size_t i = 0; i < n_args; ++i) {
+        digitalio_digitalinout_obj_t *pin = MP_OBJ_TO_PTR(args[i]);
+        if (common_hal_digitalio_digitalinout_get_direction(pin) !=
+            DIRECTION_INPUT) {
+            common_hal_digitalio_digitalinout_switch_to_input(pin, PULL_UP);
+        }
+        digitalio_pull_t pull = common_hal_digitalio_digitalinout_get_pull(pin);
+        if (pull == PULL_NONE) {
+            common_hal_digitalio_digitalinout_set_pull(pin, PULL_UP);
+        }
+        if (pull != PULL_DOWN) {
+            gamepad_singleton->pulls |= 1 << i;
+        }
+        gamepad_singleton->pins[i] = pin;
```

Comment:
Why move this into shared-bindings? I think it should be in shared-module.

Reply:
It uses less space like that, and it's much easier to debug having all the code in one place.

Reply:
Ya, I'd prefer it in shared-module so that it is consistent with everything else. How did you measure the space difference? I'd expect it to be inlined and generate the same result.

Reply:
I looked at the free flash displayed at the end of compilation, but this wasn't a scientific test, and there were several changes at once, so it's possible that particular one indeed didn't make any difference. I will reorganize it all, since I need separate files for the second class anyways.

Reply:
By the way, is there a way to tell the compiler to *not* inline a particular function? For example, the ``validate_pin`` is rather short, so it would be inlined multiple times, but we don't particularly care about the speed of initialization, that is only done once anyways, and it would be better to save that space for cases when we do need the speed...

Reply:
You can `__attribute__ ((noinline))` to the function declaration. I think there may be some cases where it ignores the advice, but you could try it with and without and see if you see a code difference. But in general we haven't tried to second-guess the compiler on inlining, except to adjust `-finline-limit`. I think `-Os` may make the decision in a relatively sophisticated way.

Reply:
You are right, in the long run it will actually be probably detrimental to mess with the algorithm.